src/effects/RealtimeEffectStateUI.cpp

```diff
@@ -0,0 +1,255 @@
+
+      // Effect has no client interface
+      if (client == nullptr)
+         return;
+
+      // mUIHost will acquire the correct instance
+      // and will manage its lifetime
+      wxWindow& projectWindow = ProjectWindow::Get(project);
+      
+      std::shared_ptr<EffectInstance> pInstance;
+
+      auto access = currentState->GetAccess();
+
+      // EffectUIHost invokes shared_from_this on access
+      Destroy_ptr<EffectUIHost> dlg(safenew EffectUIHost(
+         &projectWindow, project, *effectPlugin, *client, pInstance, *access,
+         currentState));
+
+      if (!dlg->Initialize())
+         return;
+
+      // pInstance is updated in EffectUIHost constructor  
+      mCurrentInstance = pInstance;
+      // Dialog is owned by its parent now!
+      mEffectUIHost = dlg.release();
+
+      UpdateTitle();
+
+      client->ShowClientInterface(
+         projectWindow, *mEffectUIHost, mEffectUIHost->GetValidator(), false);
+
+      // The dialog was modal? That shouldn't have happened
+      if (mEffectUIHost == nullptr || !mEffectUIHost->IsShown())
+      {
+         assert(false);
+         mEffectUIHost = {};
+      }
+   }
+
+   // This method updates the dialog title based on the track name
+   // If the dialog is not shown - UpdateTitle caches the track name to
+   // be reused when the dialog is shown.
+   void UpdateTitle(const Track* track = nullptr)
+   {
+      if (track != nullptr)
+         mTrackName = track->GetName();
+
+      // Lazily evaluate the name of the effect
+      if (mEffectName.empty())
+      {
+         auto currentState = mCurrentState.lock();
+
+         if (!currentState)
+            return;
+         
+         const auto ID = currentState->GetID();
+         const auto effectPlugin = EffectManager::Get().GetEffect(ID);
+
+         if (!effectPlugin)
+            return;
+
+         mEffectName = effectPlugin->GetDefinition().GetName();
+      }
+
+      if (mEffectUIHost == nullptr)
+         return;
+
+      const auto title = mTrackName.empty() ?
+         mEffectName :
+         /* i18n-hint: First %s is an effect name, second is a track name */
+         XO("%s - %s").Format(mEffectName, mTrackName);
+
+      mEffectUIHost->SetTitle(title);
+      mEffectUIHost->SetName(title);
+   }
+   
+   void Hide()
+   {
+      if (mEffectUIHost != nullptr)
+      {
+         // EffectUIHost calls Destroy in OnClose handler
+         mEffectUIHost->Close();
+         mEffectUIHost = {};
+      }
+   }
+
+private:
+   std::weak_ptr<RealtimeEffectState> mCurrentState;
+   // Used to check that effect instance has not changed
+   // while the dialog is shown
+   std::weak_ptr<EffectInstance> mCurrentInstance;
+   
+   wxWeakRef<EffectUIHost> mEffectUIHost;
+
+   TranslatableString mEffectName;
+   wxString mTrackName;
+};
+
+RealtimeEffectStateUI::RealtimeEffectStateUI()
+    : mSharedState(std::make_shared<RealtimeEffectStateUI::SharedState>())
+{
+}
```

Comment:
You followed one of the suggestions for changing the management of lifetimes of the shared state.  Is it all correct?

- generate noise, open sidebar, add any effect, but do NOT yet click the gear
- Do something undo-able
- undo
- click the gear
- redo
- click the gear again

This was a sequence that I feared might erroneously leave two non-modal dialogs opened connected to the same effect state.

This doesn't happen.  Yet, the reasons why it doesn't happen seem a little fragile to me.

It doesn't happen because `ProjectHistory::PopState` fires many track deletion, then addition, events in these lines:

```
   dstTracks.Clear();
   for (auto t : tracks->Any())
      dstTracks.Add(t->Duplicate());
```

And the Subscription you added calls `HideRealtimeUIForTrack` -- with the consequence, that ANY undo or redo makes the non-modal dialog disappear, I think? -- which might be considered unacceptable to QA.

Now, the "lazy" creation of `RealtimeEffectStateUI` almost always, whenever you cause a push or pop of the Undo stack -- can mean there is little point in avoiding doing it sooner, eagerly.  And the way to do that, as I suggested, is to have `RealtimeEffectState` call `mEffectStateExtensions.BuildAll()` in the constructor.  That would defeat my doubts.

Or it would also do so, if you do the other thing:  have `RealtimeEffectState` hold a `shared_ptr` to its `Extensions` and copy that pointer in the copy ctor.  Then `BuildAll` won't be needed.  And not needing `BuildAll` might mean economy in case `Extensions` ever needs to host some other attachment.